File: src/commands/update.ts
import "../bootstrap.js";
// `tokless update` — version-aware refresh.

import { runInit } from "./init.js";
import { log } from "../util/logger.js";
import { c, sym } from "../util/colors.js";
import {
  gatherVersions,
  bustVersionCache,
  semverCompare,
  type VersionInfo,
} from "../util/versions.js";
import { listTools } from "../core/registry.js";
import type { InitOptions } from "./init.js";

export async function runUpdate(opts: InitOptions = {}): Promise<number> {
  console.log("");
  console.log("  " + c.bold(c.cyan("tokless update")) + c.gray("  refresh tools to latest"));
  console.log("");

  if (opts.dryRun) {
    log.info("Dry run — would probe registries and reinstall changed tools only.");
  }

  // 1) Probe upstream.
  bustVersionCache();
  const probing = "  " + c.gray("probing upstream…");
  if (process.stdout.isTTY) process.stdout.write(probing);
  else console.log(probing);
  let versions: Record<string, VersionInfo> = {};
  try {
    versions = await gatherVersions();
  } catch {
  }
  if (process.stdout.isTTY) process.stdout.write("\r\x1b[2K");
  else console.log("");

  // 2) Print version diff table.
  const toolsToShow = listTools().map((t) => t.id);
  const changed: string[] = [];
  for (const t of toolsToShow) {
    const info = versions[t];
    const installed = info?.installed ?? c.gray("not on PATH");
    const latest = info?.latest ?? c.gray("?");
    let mark = c.gray(sym.bullet);
    let suffix = c.gray(" (pinned)");
    if (info?.installed && info.latest && semverCompare(info.installed, info.latest) < 0) {
      mark = c.yellow("↑");
      suffix = c.yellow(" → upgrade");
      changed.push(t);
    } else if (!info?.installed && info?.latest) {
      mark = c.yellow("+");
      suffix = c.yellow(" → install");
      changed.push(t);
    } else if (info?.installed && info?.latest) {
      mark = c.green(sym.check);
      suffix = c.gray(" (up to date)");
    }
    console.log(
      `  ${mark} ${t.padEnd(14)} ${String(installed).padEnd(10)} → ${String(latest).padEnd(10)}${suffix}`,
    );
  }
  console.log("");

  if (opts.dryRun) {
    log.info(
      changed.length > 0
        ? `Would upgrade: ${changed.join(", ")}`
        : "Everything up to date.",
    );
    console.log("");
    return 0;
  }

  // 3) If nothing changed, stop — no reason to re-run the install pass.
  if (changed.length === 0) {
    log.ok("Everything up to date.");
    console.log("");
    return 0;
  }

  // 4) Re-install only the tools whose version moved.
  console.log("  " + c.bold(`Upgrading: ${changed.join(", ")}`));
  return await runInit({
    ...opts,
    yes: true,
    upgrade: true,
    tools: opts.tools ?? (changed as InitOptions["tools"]),
  });
}

File: internal/commands/update.go
package commands

import (
	"fmt"

	"github.com/HoangP8/tokless/internal/agents"
	"github.com/HoangP8/tokless/internal/core"

	toolsPkg "github.com/HoangP8/tokless/internal/tools"
	"github.com/HoangP8/tokless/internal/util"
)

func RunUpdate(opts InitOptions) int {
	MaybeSelfUpdate(opts)

	cmdHeader("update", "refresh tools to latest")

	if opts.DryRun {
		util.L.Raw("  " + util.C.Cyan(util.Sym.Info) + " Dry run — probe only, no installs.")
		util.L.Raw("")
	}

	probingLine := "  " + util.C.Gray("probing upstream…")
	if stdoutTTY() {
		fmt.Print(probingLine)
	} else {
		util.L.Raw(probingLine)
	}
	versions := util.GatherVersionsForce()
	if stdoutTTY() {
		fmt.Print(util.EraseStyledLine(probingLine))
	} else {
		util.L.Raw("")
	}

	var changed []string
	util.TreeTop("Versions")
	for _, t := range core.ListTools() {
		if t.InstructionOnly {
			continue
		}
		info, has := versions[t.ID]
		name := paintName(padEnd(t.ID, 14))

		installed := util.C.Dim("not on PATH")
		if has && info.Installed != nil {
			installed = paintVer(padEnd("v"+*info.Installed, 10))
		} else {
			installed = util.C.Dim(padEnd("not on PATH", 10))
		}

		latest := util.C.Dim(padEnd("?", 10))
		if has && info.Latest != nil {
			latest = paintVer(padEnd("v"+*info.Latest, 10))
		}

		mark := util.C.Gray(util.Sym.Bullet)
		suffix := util.C.Dim(" (latest unknown)")

		switch {
		case has && info.Installed != nil && info.Latest != nil && util.SemverCompare(info.Installed, info.Latest) < 0:
			mark = util.C.Yellow("↑")
			latest = util.C.Bold(util.C.Green(padEnd("v"+*info.Latest, 10)))
			suffix = util.C.Yellow(" → upgrade")
			changed = append(changed, t.ID)
		case has && info.Installed == nil && info.Latest != nil:
			mark = util.C.Yellow("+")
			latest = util.C.Bold(util.C.Green(padEnd("v"+*info.Latest, 10)))
			suffix = util.C.Yellow(" → install")
			changed = append(changed, t.ID)
		case has && info.Installed != nil && info.Latest != nil:
			mark = util.C.Green(util.Sym.Check)
			suffix = util.C.Green(" (up to date)")
		}

		util.TreeLeaf(mark + " " + name + " " + installed + " " + paintArrow() + " " + latest + suffix)
	}
	util.TreeClose()

	if opts.DryRun {
		if len(changed) > 0 {
			treeStatus(statusInfo(util.C.Gray("Would upgrade: ") + util.C.Bold(paintVer(joinComma(changed)))))
		} else {
			treeStatus(statusOK("Everything up to date."))
		}
		printRepoFooter(true)
		util.L.Raw("")
		return 0
	}

	if len(changed) == 0 {
		agents.PiUpdatePackages()
		treeStatus(statusOK("Everything up to date."))
		printRepoFooter(true)
		util.L.Raw("")
		return 0
	}
	if !opts.Yes && util.IsInteractive() {
		var pick []util.MultiSelectOption
		for _, t := range core.ListTools() {
			if !contains(changed, t.ID) {
				continue
			}
			info := versions[t.ID]
			installed := "not on PATH"
			if info.Installed != nil {
				installed = "v" + *info.Installed
			}
			latest := "?"
			if info.Latest != nil {
				latest = "v" + *info.Latest
			}
			hint := "install"
			if info.Installed != nil {
				hint = "upgrade"
			}
			pick = append(pick, util.MultiSelectOption{Value: t.ID, Label: padEnd(t.ID, 14) + installed + " → " + latest, Hint: hint, Selected: true})
		}
		changed = util.MultiSelect("Select tools to update", pick)
		if len(changed) == 0 {
			treeStatus(statusInfo(util.C.Dim("No tools selected.")))
			printRepoFooter(true)
			util.L.Raw("")
			return 0
		}
	}

	if !opts.DryRun {
		needNode, needGit, minNode := false, false, 0
		for _, t := range core.ListTools() {
			if contains(changed, t.ID) {
				needNode = needNode || toolNeedsNode(t)
				needGit = needGit || t.NeedsGit
				if t.MinNodeMajor > minNode {
					minNode = t.MinNodeMajor
				}
			}
		}
		nodeOK, gitOK := util.EnsureDeps(needNode, needGit, minNode)
		if !nodeOK || !gitOK {
			var keep []string
			for _, id := range changed {
				tool := core.GetTool(id)
				if tool == nil {
					continue
				}
				if toolNeedsNode(tool) && !nodeOK {
					continue
				}
				if tool.NeedsGit && !gitOK {
					continue
				}
				keep = append(keep, id)
			}
			changed = keep
		}
		if len(changed) == 0 {
			treeStatus(util.C.Red(util.Sym.Cross) + " " + util.C.Gray("Missing dependencies; nothing safe to update."))
			printRepoFooter(true)
			util.L.Raw("")
			return 1
		}
	}
	allTools := core.ListTools()
	var tools []*core.ToolManifest
	for _, t := range allTools {
		if contains(changed, t.ID) {
			tools = append(tools, t)
		}
	}
	bar := util.NewSectionProgress("Upgrading " + joinComma(changed))
	bar.Start(len(tools))
	for _, tool := range tools {
		bar.Begin(tool.Label)
		report := func(phase string, frac float64) { bar.Step(phase, frac) }
		err := util.WithSilencedLogs(func() error {
			_, e := tool.Install(core.RunOpts{DryRun: opts.DryRun, Upgrade: true, Report: report})
			return e
		})
		if err != nil {
			bar.Fail(firstLine(err.Error()))
		} else {
			bar.Complete("")
		}
	}
	bar.Done("")

	// Re-pin upgraded tools' per-agent config to the freshly installed version.
	if !opts.DryRun {
		toolsPkg.ConfigureInstructionConflicts(true)
		resyncWiring(tools)
		toolsPkg.ConfigureInstructionConflicts(false)
		agents.PiUpdatePackages()
	}

	// Upgrade mutated installed versions; drop cached latest so next read is fresh.
	if !opts.DryRun {
		util.BustVersionCache()
	}
	treeStatus(statusOK("Updated " + joinComma(changed) + "."))
	printRepoFooter(true)
	util.L.Raw("")
	return 0
}

// resyncWiring re-runs WireFor for each upgraded tool only on agents where it is
// already wired (gated by VerifyFor), syncing version pins without newly wiring.
func resyncWiring(tools []*core.ToolManifest) {
	agents := core.ListAgents()
	for _, tool := range tools {
		for _, agent := range agents {
			wire, ok := tool.WireFor[agent.ID]
			if !ok || !agent.Detect().Installed {
				continue
			}
			if verify, vok := tool.VerifyFor[agent.ID]; vok {
				if r := verify(); r == nil || !*r {
					continue
				}
			}
			_ = util.WithSilencedLogs(func() error {
				_, e := wire(core.RunOpts{Upgrade: true})
				return e
			})
		}
	}
}
